### DISCUSSION TITLE
Runtime Interface Type Validation

### DISCUSSION THREAD

**Discussion by shawnmuzick** [prisma/prisma#17080] [Q&A]
I have 2 routes in an express app, for put and patch, for altering full objects and altering partial objects, like such:
```
//should be a whole object, not just the changed data
router.put(
  "/users/:id",
  asyncHandler(async (req: Request, res: Response) => {
    let params = { where: { pk_id: req.params.id }, data: req.body.data };
    //need to check that data fully implements the user interface, i.e.
    /*
    user{
       name:string;
       email:string;
       phone:string;
       //etc
    }
    */
    let result = await prisma.users.update(params);
    res.send({ data: result });
  })
);
//does not need to be a whole object, just the fields that change
router.patch(
  "/users/:id",
  asyncHandler(async (req: Request, res: Response) => {
    let params = { where: { pk_id: req.params.id }, data: req.body.data };
    //data does not need to fully implement the user interface
    let result = await prisma.users.update(params);
    res.send({ data: result });
  })
);
```
The types exposed by Prisma are interface types, rather than classes, and so can't be checked against an object via instanceof. The underlying javascript has no concept of interfaces. Most solutions to this I have seen on other forums suggest manually checking for a particular property or method on the object.
```
interface A {
    member: string;
}

function instanceOfA(object: any): object is A {
    return 'member' in object;
}

var a: any = {member: "foobar"};

if (instanceOfA(a)) {
    alert(a.member);
}
````

While these solutions work, they break the "single source of truth", as they're manually defined, independent of changes to the schema.  

How can I check that the shape of the data I receive at runtime matches all fields of a model, without manually implementing the interface types in classes or manually checking particular fields?

**Comment by nurul3101** [comment]
Hello @shawnmuzick 👋 

You can use validator libraries like [Joi](https://joi.dev/) or [Zod](https://zod.dev/) to ensure that data is in the expected format.

But I believe you would like to use the default generated types by Prisma based on your models and avoid using these external libraries?

Using these libraries would help you in further validating the input your receive, example in case you want to validate if the string passed is a valid email, you could use the inbuilt methods like `.email()` to validate.

Do the generated prisma types work as expected when you use them in the `PUT` API call? And you face issue only when implementing the `PATCH` API call as all fields of User model are not needed?

**Comment by shawnmuzick** [comment]
I guess the issue more generally is that: I can write my own validation, or use a library, but I doesn't make sense that I'd need to manually define objects/classes when Prisma introspects the database and generates them for me. The types all work, it's just that I need to manually define what I need outside of what prisma already provides, which seems to miss the point.
The author of this package seems to echo my sentiment: https://github.com/CarterGrimmeisen/zod-prisma. In his own words "I got tired of having to manually create Zod schemas for my Prisma models and of updating them everytime I made schema changes. This provides a way of automatically generating them with your prisma"

**Reply by nurul3101** [reply]
I get what you are saying, we have a related feature request here, if you can give a 👍 to the feature request then it would help our team in prioritising it.
- https://github.com/prisma/prisma/issues/3528

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
